### DISCUSSION TITLE
Connection Timeout, Socket Timeout, Read Timeout의 차이점

### DISCUSSION THREAD

**Discussion by le2sky** [2024-woowacourse-study/level-interview#144] [Network]
뭘까용

**Comment by alstn113** [comment]
## Connection Timeout

Connection Timeout은 클라이언트가 서버에 연결을 시도할 때, 일정 시간 내에 연결이 이루어지지 않으면 발생하는 타임아웃입니다.
TCP 소켓 통신에서 클라이언트와 서버가 연결될 때, 정확한 전송을 보장하기 위해 사전에 세션을 수립하는데, 이 과정을 3-way-handshake라고 합니다.
Connection Timeout은 이 3-way-handshake가 일정 시간 내에 완료되지 않을 때 발생합니다.
즉, 서버의 장애나 응답 지연으로 인해 연결을 맺지 못하면 Connection Timeout이 발생합니다.

## Socket Timeout / Read Timeout

### Socket Timeout

Socket Timeout은 Connection Timeout 이후에 발생할 수 있는 타임아웃입니다. 
클라이언트와 서버가 연결된 후, 서버는 데이터를 클라이언트에게 전송합니다. 
이 때 하나의 데이터 덩어리가 아니라 여러 개의 패킷 단위로 쪼개서 전송되는데, 각 패킷이 전송될 때의 시간 차이 제한을 Socket Timeout이라고 합니다. 
만약 서버가 일정 시간 내에 다음 패킷을 보내지 않으면, 클라이언트는 Socket Timeout을 발생시키고 연결을 종료할 수 있습니다.

### Read Timeout

Read Timeout은 클라이언트와 서버가 연결된 후, 특정 I/O 작업이 일정 시간 내에 완료되지 않으면 발생하는 타임아웃입니다. 
클라이언트와 서버가 연결된 상태에서, 서버의 응답이 지연되거나 I/O 작업이 길어져 요청이 처리되지 않을 때 클라이언트는 연결을 끊습니다. 
Read Timeout은 이러한 상황을 방지하기 위해 설정된 타임아웃으로, 일정 시간 내에 데이터가 읽혀지지 않으면 클라이언트가 연결을 종료합니다.

**Comment by 3Juhwan** [comment]
**connect timeout**

클라이언트가 서버에 연결을 시도할 때, 연결이 성공하기까지 기다리는 최대 시간을 의미합니다. 서버와의 연결은 TCP handshake를 의미합니다. 설정한 시간 내에 서버에 연결을 성공하지 못하면 connection timeout이 발생합니다. 

- 호스트 서버에서 지정된 포트를 수신하는 서비스가 없는 경우
- 호스트 서버가 연결을 허용하지 않는 경우
- 호스트 서버에 문제가 있어 서비스할 수 없는 경우
- 네트워크 문제가 있어 연결이 불가능한 경우

**read timeout**

소켓에서 데이터를 읽어들일 때, 데이터 읽기가 완료되기까지 기다리는 최대 시간을 의미합니다. 설정한 시간 내에 소켓으로부터 1바이트의 데이터라도 읽지 못하면 read timeout이 발생합니다. 

- 서버와 연결은 되었는데, 응답 전송이 지연되는 경우
- 서버의 요청 처리 속도가 느린 경우

**socket timeout**

서버와 TCP 연결에 성공하고 데이터를 여러 패킷으로 전달받습니다. 패킷 간의 전송 시간에 차이가 있는데, 이 차이의 최대 시간을 의미합니다. 패킷 간의 전송 시간 차이가 설정한 시간을 넘어가면 socket timeout이 발생합니다. 예를 들어, 처음 도착한 패킷과 마지막으로 도착한 패킷의 시간 차이가 임계값을 넘어가면 socket timeout이 발생합니다. socket timeout은 read timeout의 일종으로 java에서 혼용해서 사용합니다.

**Comment by J-I-H-O** [comment]
![image](https://github.com/2024-woowacourse-study/level-interview/assets/110461155/54df61de-8e1a-4e97-8e6b-25e67525be69)

![image](https://github.com/2024-woowacourse-study/level-interview/assets/110461155/8f17ff88-af14-42ce-9922-9c190c28d71e)
